prometheus-node-exporter-lua: add realtek-poe exporter

hirnpfirsich opened the pull request:
Maintainer: `Gregor Michels <hirnpfirsich@brainpeach.de>`
Compile tested: for `realtek rtl838x` on `amd64` (`snapshot`)
Run tested: `realtek rtl838x`, `Zyxel GS190010HP`, (`snapshot`)

Description:
```
exposes PoE metrics obtained from realtek-poe

 # HELP realtek_poe_switch_info information about the poe controller
 # TYPE realtek_poe_switch_info gauge
 realtek_poe_switch_info{mcu="ST Micro ST32F100 Microcontroller",firmware="v22.4"} 1

 # HELP realtek_poe_switch_budget_watts overall power budget
 # TYPE realtek_poe_switch_budget_watts gauge
 realtek_poe_switch_budget_watts 77

 # HELP realtek_poe_switch_consumption_watts overall power consumption
 # TYPE realtek_poe_switch_consumption_watts gauge
 realtek_poe_switch_consumption_watts 5

 # HELP realtek_poe_port_priority poe priority of port
 # TYPE realtek_poe_port_priority gauge
 realtek_poe_port_priority{device="lan1"} 1
 realtek_poe_port_priority{device="lan2"} 1
 [...]

 # HELP realtek_poe_port_consumption_watts per port power consumption
 # TYPE realtek_poe_port_consumption_watts gauge
 realtek_poe_port_consumption_watts{device="lan1"} 0
 realtek_poe_port_consumption_watts{device="lan2"} 0
 [...]

 # HELP realtek_poe_port_state per port poe state
 # TYPE realtek_poe_port_state gauge
 realtek_poe_port_state{device="lan1",state="Disabled"} 0
 realtek_poe_port_state{device="lan1",state="Searching"} 1
 [...]
 (states: Disabled, Searching, Delivering power, Fault, Other fault, Requesting power)

 # HELP realtek_poe_port_mode per port poe mode
 # TYPE realtek_poe_port_mode gauge
 realtek_poe_port_mode{device="lan1",mode="PoE"} 0
 realtek_poe_port_mode{device="lan1",mode="PoE+"} 1
 [...]
 (modes: PoE, Legacy, pre-PoE+, PoE+)

Signed-off-by: Gregor Michels <hirnpfirsich@brainpeach.de>
```

hirnpfirsich commented:
Hopefully this exporter is useful for some people :)

We ([Freifunk Leipzig](https://freifunk-leipzig.de)) needed this exporter to monitor the PoE status of Zyxel GS1900-10HP switches.
It has been running over the past month without issues on two of theses switches.
Since `realtek-poe` finally landed upstream I thought I could give upstreaming this a shot.

champtar commented:
Hi @hirnpfirsich 
Shouldn't `realtek_poe_port_state_*` and `realtek_poe_port_mode_*` be `realtek_poe_port_info{port=,mode=,state=}` ? This would remove 8 metrics per port, so 80 metrics (https://www.robustperception.io/exposing-the-software-version-to-prometheus/)
Also maybe use `realtek_poe_switch_*` to make it clearer, and finally the convention is to suffix with the unit, so `_watts` (https://prometheus.io/docs/practices/naming/)

champtar commented on utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/realtek-poe.lua:
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env lua
```
```suggestion
```
This is not needed

hirnpfirsich replied:
Oh right because `prometheus-node-exporter-lua` actually imports the collectors as libraries (or the `lua` equivalent of that) instead of executing it (which totally makes sense because I have not imported anything related to prometheus but can easily call `metric`).

Thanks for the hint :)

hirnpfirsich commented:
> Hi @hirnpfirsich Shouldn't `realtek_poe_port_state_*` and `realtek_poe_port_mode_*` be `realtek_poe_port_info{port=,mode=,state=}` ? This would remove 8 metrics per port, so 80 metrics

Hello @champtar, thank you for you're timely response :)
I totally agree that a unique metric for every state is overkill. Therefore I refactored these into two metrics:
* `realtek_poe_switch_port_mode` and
* `realtek_poe_switch_port_state`

which expand by labels with every mode and state like so:
```
realtek_poe_switch_port_mode{device="lan1",mode="legacy"} 1
realtek_poe_switch_port_mode{device="lan1",mode="poe"} 0
...
realtek_poe_switch_port_state{device="lan1",state="disabled"} 0
realtek_poe_switch_port_state{device="lan1",state="searching"} 0
realtek_poe_switch_port_state{device="lan1",state="powering"} 1
...
```

This way a query against an explicit state (like `Powering`) always returns data.
Otherwise a port switching from `Powering` to `Searching` would not have data for all points in a timeseries with the query `realtek_poe_switch_port_state{state="Powering"} == 1`.
The same thing applies to the `realtek_poe_switch_port_mode`.

> Also maybe use `realtek_poe_switch_*` to make it clearer, and finally the convention is to suffix with the unit, so `_watts` (https://prometheus.io/docs/practices/naming/)

done and done :)

To remove the redundant information about the overall power consumption of the switch I removed the `realtek_poe_consumption` metric.

And last but not least I added a metric exposing the mcu and firmware version of the switch: `realtek_poe_switch_info`.

An updated and detailed list of all metrics is now the new commit message.

champtar commented:
> > Hi @hirnpfirsich Shouldn't `realtek_poe_port_state_*` and `realtek_poe_port_mode_*` be `realtek_poe_port_info{port=,mode=,state=}` ? This would remove 8 metrics per port, so 80 metrics
> 
> Hello @champtar, thank you for you're timely response :) I totally agree that a unique metric for every state is overkill. Therefore I refactored these into two metrics:
> 
>     * `realtek_poe_switch_port_mode` and
> 
>     * `realtek_poe_switch_port_state`
> 
> 
> which expand by labels with every mode and state like so:
> 
> ```
> realtek_poe_switch_port_mode{device="lan1",mode="legacy"} 1
> realtek_poe_switch_port_mode{device="lan1",mode="poe"} 0
> ...
> realtek_poe_switch_port_state{device="lan1",state="disabled"} 0
> realtek_poe_switch_port_state{device="lan1",state="searching"} 0
> realtek_poe_switch_port_state{device="lan1",state="powering"} 1
> ...
> ```
> 
> This way a query against an explicit state (like `Powering`) always returns data. Otherwise a port switching from `Powering` to `Searching` would not have data for all points in a timeseries with the query `realtek_poe_switch_port_state{state="Powering"} == 1`. The same thing applies to the `realtek_poe_switch_port_mode`.

Googling a bit what you did seems to be the normal way to represent enum with prometheus.
Something like `realtek_poe_port_state{device="lan1"} 2` for `Delivering power` would be more efficient but less usable (manual mapping in grafana ?)
 
> > Also maybe use `realtek_poe_switch_*` to make it clearer, and finally the convention is to suffix with the unit, so `_watts` (https://prometheus.io/docs/practices/naming/)
> 
> done and done :)

Sorry for the confusion, I meant `realtek_poe_port_*` and `realtek_poe_switch_*` :)

> To remove the redundant information about the overall power consumption of the switch I removed the `realtek_poe_consumption` metric.

Is it really redundant ? ie is it only the sum of the port consumption ?

> And last but not least I added a metric exposing the mcu and firmware version of the switch: `realtek_poe_switch_info`.

nice

> An updated and detailed list of all metrics is now the new commit message.

don't forget to update the PR description

hirnpfirsich commented:
> Googling a bit what you did seems to be the normal way to represent enum with prometheus.
> Something like `realtek_poe_port_state{device="lan1"} 2` for `Delivering power` would be more efficient but less usable (manual mapping in grafana ?)

I would like stick to the less efficient version, because currently there is no support for `# HELP` strings in `metric()`. Mapping in grafana is certainly possible but not every interaction with prometheus goes through a grafana instance. 

> Sorry for the confusion, I meant `realtek_poe_port_*` and `realtek_poe_switch_*` :)

That is prettier indeed ! done :)

> Is it really redundant ? ie is it only the sum of the port consumption ?

That is a good question. After reading through the [realtek-poe - main.c](https://github.com/Hurricos/realtek-poe/blob/realtek-poe/src/main.c) it seems to me that the overall consumption and per port consumption are not calculated from each other. Instead these values come from different commands/replies from the `mcu`. These two functions are the callbacks for these commands/replies:
```c
static int
poe_reply_power_stats(unsigned char *reply)
{
	state.power_consumption = read16_be(reply + 2) * 0.1;

	return 0;
}
```
and
```c
static int
poe_reply_port_power_stats(unsigned char *reply)
{
	int port_idx = reply[2];

	state.ports[port_idx].watt = read16_be(reply + 9) * 0.1;
	return 0;
}
```
Maybe @mrnuke or @Hurricos could ack that ? :)
Therefore I added `realtek_poe_switch_consumption_watts` back in.

hirnpfirsich commented:
Please do not merge this PR yet @champtar. I'am currently not able to test the new changes on a switch (`201457c17ac03a821974749028e4e30bfcee6593`). Will do that later this day and notify you

hirnpfirsich commented:
> don't forget to update the PR description

Updated the description

champtar commented:
> Please do not merge this PR yet @champtar. I'am currently not able to test the new changes on a switch (`201457c17ac03a821974749028e4e30bfcee6593`). Will do that later this day and notify you

Converted to draft

mrnuke commented:
Per-port and total numbers come from different sources, yes. Sum(ports) should equal total, but realtek-poe doesn't guarantee that.

hirnpfirsich commented:
> Per-port and total numbers come from different sources, yes. Sum(ports) should equal total, but realtek-poe doesn't guarantee that.

thanks, appreciate it :)

hirnpfirsich commented:
@champtar successfully tested the latest commit (`201457c17`)

```
$ curl -s http://192.168.55.185:9100/metrics | grep realtek_poe > metrics
$ cat metrics && promtool check metrics < metrics 
# TYPE realtek_poe_switch_info gauge
realtek_poe_switch_info{mcu="ST Micro ST32F100 Microcontroller",firmware="v22.4"} 1
# TYPE realtek_poe_switch_budget_watts gauge
realtek_poe_switch_budget_watts 77
# TYPE realtek_poe_switch_consumption_watts gauge
realtek_poe_switch_consumption_watts 3.7999999523163
# TYPE realtek_poe_port_priority gauge
realtek_poe_port_priority{device="lan6"} 1
realtek_poe_port_priority{device="lan1"} 1
realtek_poe_port_priority{device="lan2"} 1
realtek_poe_port_priority{device="lan5"} 1
realtek_poe_port_priority{device="lan8"} 1
realtek_poe_port_priority{device="lan3"} 1
realtek_poe_port_priority{device="lan4"} 1
realtek_poe_port_priority{device="lan7"} 1
# TYPE realtek_poe_port_consumption_watts gauge
realtek_poe_port_consumption_watts{device="lan6"} 0
realtek_poe_port_consumption_watts{device="lan1"} 0
realtek_poe_port_consumption_watts{device="lan2"} 0
realtek_poe_port_consumption_watts{device="lan5"} 0
realtek_poe_port_consumption_watts{device="lan8"} 0
realtek_poe_port_consumption_watts{device="lan3"} 3.7000000476837
realtek_poe_port_consumption_watts{device="lan4"} 0
realtek_poe_port_consumption_watts{device="lan7"} 0
# TYPE realtek_poe_port_state gauge
realtek_poe_port_state{device="lan6",state="Disabled"} 0
realtek_poe_port_state{device="lan1",state="Disabled"} 0
realtek_poe_port_state{device="lan2",state="Disabled"} 0
realtek_poe_port_state{device="lan5",state="Disabled"} 0
realtek_poe_port_state{device="lan8",state="Disabled"} 0
realtek_poe_port_state{device="lan3",state="Disabled"} 0
realtek_poe_port_state{device="lan4",state="Disabled"} 0
realtek_poe_port_state{device="lan7",state="Disabled"} 0
realtek_poe_port_state{device="lan6",state="Searching"} 1
realtek_poe_port_state{device="lan1",state="Searching"} 1
realtek_poe_port_state{device="lan2",state="Searching"} 1
realtek_poe_port_state{device="lan5",state="Searching"} 1
realtek_poe_port_state{device="lan8",state="Searching"} 1
realtek_poe_port_state{device="lan3",state="Searching"} 0
realtek_poe_port_state{device="lan4",state="Searching"} 1
realtek_poe_port_state{device="lan7",state="Searching"} 1
realtek_poe_port_state{device="lan6",state="Delivering power"} 0
realtek_poe_port_state{device="lan1",state="Delivering power"} 0
realtek_poe_port_state{device="lan2",state="Delivering power"} 0
realtek_poe_port_state{device="lan5",state="Delivering power"} 0
realtek_poe_port_state{device="lan8",state="Delivering power"} 0
realtek_poe_port_state{device="lan3",state="Delivering power"} 1
realtek_poe_port_state{device="lan4",state="Delivering power"} 0
realtek_poe_port_state{device="lan7",state="Delivering power"} 0
realtek_poe_port_state{device="lan6",state="Fault"} 0
realtek_poe_port_state{device="lan1",state="Fault"} 0
realtek_poe_port_state{device="lan2",state="Fault"} 0
realtek_poe_port_state{device="lan5",state="Fault"} 0
realtek_poe_port_state{device="lan8",state="Fault"} 0
realtek_poe_port_state{device="lan3",state="Fault"} 0
realtek_poe_port_state{device="lan4",state="Fault"} 0
realtek_poe_port_state{device="lan7",state="Fault"} 0
realtek_poe_port_state{device="lan6",state="Other fault"} 0
realtek_poe_port_state{device="lan1",state="Other fault"} 0
realtek_poe_port_state{device="lan2",state="Other fault"} 0
realtek_poe_port_state{device="lan5",state="Other fault"} 0
realtek_poe_port_state{device="lan8",state="Other fault"} 0
realtek_poe_port_state{device="lan3",state="Other fault"} 0
realtek_poe_port_state{device="lan4",state="Other fault"} 0
realtek_poe_port_state{device="lan7",state="Other fault"} 0
realtek_poe_port_state{device="lan6",state="Requesting power"} 0
realtek_poe_port_state{device="lan1",state="Requesting power"} 0
realtek_poe_port_state{device="lan2",state="Requesting power"} 0
realtek_poe_port_state{device="lan5",state="Requesting power"} 0
realtek_poe_port_state{device="lan8",state="Requesting power"} 0
realtek_poe_port_state{device="lan3",state="Requesting power"} 0
realtek_poe_port_state{device="lan4",state="Requesting power"} 0
realtek_poe_port_state{device="lan7",state="Requesting power"} 0
# TYPE realtek_poe_port_mode gauge
realtek_poe_port_mode{device="lan6",mode="PoE"} 0
realtek_poe_port_mode{device="lan1",mode="PoE"} 0
realtek_poe_port_mode{device="lan2",mode="PoE"} 0
realtek_poe_port_mode{device="lan5",mode="PoE"} 0
realtek_poe_port_mode{device="lan8",mode="PoE"} 0
realtek_poe_port_mode{device="lan3",mode="PoE"} 0
realtek_poe_port_mode{device="lan4",mode="PoE"} 0
realtek_poe_port_mode{device="lan7",mode="PoE"} 0
realtek_poe_port_mode{device="lan6",mode="Legacy"} 0
realtek_poe_port_mode{device="lan1",mode="Legacy"} 0
realtek_poe_port_mode{device="lan2",mode="Legacy"} 0
realtek_poe_port_mode{device="lan5",mode="Legacy"} 0
realtek_poe_port_mode{device="lan8",mode="Legacy"} 0
realtek_poe_port_mode{device="lan3",mode="Legacy"} 0
realtek_poe_port_mode{device="lan4",mode="Legacy"} 0
realtek_poe_port_mode{device="lan7",mode="Legacy"} 0
realtek_poe_port_mode{device="lan6",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan1",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan2",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan5",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan8",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan3",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan4",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan7",mode="pre-PoE+"} 0
realtek_poe_port_mode{device="lan6",mode="PoE+"} 1
realtek_poe_port_mode{device="lan1",mode="PoE+"} 1
realtek_poe_port_mode{device="lan2",mode="PoE+"} 1
realtek_poe_port_mode{device="lan5",mode="PoE+"} 1
realtek_poe_port_mode{device="lan8",mode="PoE+"} 1
realtek_poe_port_mode{device="lan3",mode="PoE+"} 1
realtek_poe_port_mode{device="lan4",mode="PoE+"} 1
realtek_poe_port_mode{device="lan7",mode="PoE+"} 1
realtek_poe_port_consumption_watts no help text
realtek_poe_port_mode no help text
realtek_poe_port_priority no help text
realtek_poe_port_state no help text
realtek_poe_switch_budget_watts no help text
realtek_poe_switch_consumption_watts no help text
realtek_poe_switch_info no help text
```
